js/data/program_configuration.js

```diff
@@ -22,227 +20,204 @@ const assert = require('assert');
  * @private
  */
 class ProgramConfiguration {
+
+    constructor() {
+        this.attributes = [];
+        this.uniforms = [];
+        this.interpolationUniforms = [];
+        this.pragmas = {vertex: {}, fragment: {}};
+        this.cacheKey = '';
+    }
+
     static createDynamic(attributes, layer, zoom) {
         const self = new ProgramConfiguration();
 
-        self.attributes = [];
-        self.uniforms = [];
-        self.vertexPragmas = { define: {}, initialize: {} };
-        self.fragmentPragmas = { define: {}, initialize: {} };
+        for (const attributeConfig of attributes) {
+            const attribute = normalizePaintAttribute(attributeConfig, layer);
+            const name = attribute.name.slice(2);
 
-        const fragmentInit = self.fragmentPragmas.initialize;
-        const fragmentDefine = self.fragmentPragmas.define;
-        const vertexInit = self.vertexPragmas.initialize;
-        const vertexDefine = self.vertexPragmas.define;
+            if (layer.isPaintValueFeatureConstant(attribute.property)) {
+                self.addZoomDrivenAttribute(name, attribute);
+            } else if (layer.isPaintValueZoomConstant(attribute.property)) {
+                self.addDataDrivenAttribute(name, attribute);
+            } else {
+                self.addDataAndZoomDrivenAttribute(name, attribute, layer, zoom);
```

Comment:
In the codebase we use the terminology `property functions`, `zoom functions`, and `zoom and property functions`."data-driven" is a broader designation that may include unrelated features in the future.